### PR TITLE
perf(memory): 嵌入向量改成 base64+int8 量化 + recall 粗排批量化

### PR DESCRIPTION
**Pull request by wehos**
## Summary

- **存储**：`embedding` 字段从 `list[float]` 改成 base64(`fp16_scale | int8_dims`)；per-vector scale 保留 L2 归一化向量精度。256d 实测 5607B → 346B（~16x），余弦误差 < 0.003%
- **召回粗排**：`recall._coarse_rank` 改成把候选 + query 各 decode 一次堆成 numpy 矩阵，`(N,D)@(D,Q)` 一次乘出全部 cosine 再 max-pool。N=5000 × D=256 时从 Python 循环的 ~500ms 降到 BLAS 的毫秒级
- **迁移**：`is_cached_embedding_valid` 拒绝旧 `list[float]` 形态，worker 下一轮 sweep 自然 re-stamp。线上 ~50 条历史数据一次性成本可忽略

## Background

Issue 出在每条 256d float 向量序列化为 JSON 要 ~5.3 KB（每个 float 取 ~21 字节），加上 Python list 内存对象 28 B/float。小八的 facts.json 现在已经是 60% 体积都在向量上；warmup 跑满后 facts.json 估到 ~240 KB，每加一条 fact 整文件重写。同时 `cosine_similarity` 是纯 Python 求和循环，5k 条 × 256d 单查询 ~500ms，是 100x scale 的真正阻碍。

## Test plan

- [x] `tests/unit/test_embedding_service.py` 全部通过（含新增 round-trip / legacy-list 失效 / decode 入口测试）
- [x] `tests/unit/test_embedding_worker.py` 通过（cache-hit fixture 改用 base64，stamp 输出用 decode 比对，int8 容差 0.02）
- [x] `tests/unit/test_memory_recall.py` 通过（`_obs(embedding=...)` 自动 encode）
- [x] `tests/unit/test_fact_dedup.py` 通过（零改动；走 cosine_similarity 反向兼容路径）
- [x] memory/recall/persona/reflection 全 unit test 共 334 个 ✅

## 改动文件

- [memory/embeddings.py](memory/embeddings.py) — 编解码 helper + stamp/is_cached/cosine_similarity 适配
- [memory/recall.py](memory/recall.py) — 批量化 `_coarse_rank`
- 测试 fixture 三处适配

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## 发布说明

* **重构**
  * 嵌入向量改为紧凑的 base64+fp16 存储并统一编码/解码接口。
  * 检索相似度计算改为向量化矩阵运算，提升批量检索性能与稳定性。
  * 嵌入缓存校验更严格，自动拒绝旧格式或维度不匹配的条目。

* **修复**
  * 相似度接口对缺失/空/维度不匹配或损坏向量返回 0.0，行为更稳健且可预测。

* **测试**
  * 单元测试覆盖增强，验证新持久化格式、解码容错、维度解析与重写行为。

### CODE REVIEW

File: memory/embeddings.py [hunk 1]
```diff
@@ -844,17 +970,23 @@ def is_cached_embedding_valid(
...
+
     Any mismatch → False, callers should clear the embedding fields and
     re-enqueue the entry for the warmup worker.
     """
     if not isinstance(entry, dict):
         return False
     emb = entry.get("embedding")
-    if not isinstance(emb, list) or not emb:
+    if not isinstance(emb, str) or not emb:
         return False
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/embeddings.py:1025]: Fixed in 08b63e9b1.

`is_cached_embedding_valid` now decodes the payload and verifies the resulting length matches the dim parsed out of `model_id` (`build_model_id` always emits `<profile>-<dim>d-<quant>`, so it's authoritative). Corrupt base64 → `decoded is None` → invalid. Truncated payload → length mismatch → invalid. Both cases now force a worker re-stamp instead of sticking forever.

Regression test: `test_is_cached_embedding_valid_corrupt_base64_invalidates` and `test_is_cached_embedding_valid_wrong_dim_payload_invalidates` in `tests/unit/test_embedding_service.py`.

File: memory/recall.py [hunk 2]
```diff
@@ -276,24 +276,68 @@ async def _coarse_rank(
...
+            # decode (corrupt base64, or a future format we don't
+            # know) falls through to the unembedded pool rather than
+            # crashing the rerank.
+            if cvec is None or cvec.size == 0:
+                unembedded.append(o)
+                continue
+            if target_dim is None:
+                target_dim = int(cvec.size)
+            elif cvec.size != target_dim:
+                # Mixed dims under a single model_id should be
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/recall.py:326]: Fixed in 08b63e9b1.

`_coarse_rank` now derives `target_dim` from the running service's `model_id` (via the new `parse_dim_from_model_id`), not from whichever candidate decoded first. `model_id` encodes dim by construction, so any candidate whose decoded length disagrees is the bad one and goes to the unembedded pool — order-independent. Falls back to first-row dim only if the id is unparseable.

Note that the strengthened `is_cached_embedding_valid` (P1 fix in the same commit) already filters wrong-dim rows under a parseable `model_id` before they reach the matrix builder, so this is also defense-in-depth.

Regression test: `test_coarse_rank_first_wrong_dim_does_not_poison_others` in `tests/unit/test_memory_recall.py`.

**Comment by wehos** [comment]
@coderabbitai 已在 08b63e9b1 处理。`fact-old` 和 `fact-a` 都改用 `_encode_vector_int8(base_emb)` + 匹配的 `_embedding_text_sha256(text)` 占位，让它们走真正的 cache-hit 路径，dedup live-getter 测试现在严格对应 "old/alpha 已有 valid embedding，new/b 这次 sweep 才被 embed" 的原始意图，不再依赖 worker 顺路 re-embed legacy list 形态来歪打正着喵～

顺带：同 commit 还吃掉了 Codex 的 P1（is_cached_embedding_valid 加 decode + dim sanity）和 P2（_coarse_rank 用 model_id 推 target_dim），也是这轮 review 一起处理的。

File: memory/embeddings.py [hunk 3]
```diff
@@ -104,6 +137,118 @@ class _DisableReason(enum.Enum):
...
+    P2). We parse it here so :func:`is_cached_embedding_valid` and
+    :func:`MemoryRecallReranker._coarse_rank` can both validate
+    decoded vector shape against the *expected* dim — without this,
+    corrupt or wrong-length payloads would still pass the typeof
+    check and either stick forever (worker won't re-stamp) or poison
+    the batched matmul with a single bad row.
+    """
+    if not model_id or not isinstance(model_id, str):
+        return None
+    m = _MODEL_ID_DIM_RE.search(model_id)
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/embeddings.py:235]: Fixed in f8b665fb8.

`parse_dim_from_model_id` now uses an anchored regex `-(\d+)d-(?:int8|fp32)$` — picks the trailing `-<dim>d-<quant>` segment that `build_model_id` always emits, so a profile name with its own `-Nd-` segment can't shadow the runtime dim. Default `local-text-retrieval-v1` doesn't hit the bug today, but locking the invariant in test prevents a future profile rename from silently breaking the cache.

Regression test: `test_parse_dim_from_model_id_picks_runtime_dim_under_ambiguous_profile` covers `model-384d-v2-256d-int8 → 256` plus a couple of edge cases (None / empty / unknown quant suffix → None for graceful fallback).

File: memory/embeddings.py [hunk 4]
```diff
@@ -104,6 +137,118 @@ class _DisableReason(enum.Enum):
...
+        raw = base64.b64decode(encoded, validate=False)
+    except Exception:  # noqa: BLE001 — malformed base64 → treat as missing
+        return None
+    if len(raw) < 2:
+        return None
+    scale = float(np.frombuffer(raw[:2], dtype=np.float16)[0])
+    quantized = np.frombuffer(raw[2:], dtype=np.int8)
+    if quantized.size == 0:
+        return np.zeros(0, dtype=np.float32)
+    return quantized.astype(np.float32) * (scale / _QUANT_SCALE)
```
[the remark this answers is not in the document]
Comment (wehos) [reply memory/embeddings.py]: Fixed in f8b665fb8.

`_decode_vector_int8` now passes `validate=True` to `b64decode` (rejects non-alphabet bytes — the `!!!!`-suffix scenario) and gates on `np.isfinite(scale)` after reading the fp16 prefix (rejects NaN / ±Inf).

Note that the inner-byte tampering case can't be detected without a checksum, but `is_cached_embedding_valid` already verifies decoded length against `parse_dim_from_model_id`, so payloads whose tampering changes the byte count fall out at validity time.

Regression test: `test_decode_vector_int8_rejects_garbage_suffix_and_nan_scale` exercises garbage suffix, non-base64 input, NaN scale, and +Inf scale.

**Comment by wehos** [comment]
@coderabbitai 第三条（test_embedding_worker.py:356-382）也在 f8b665fb8 处理了：`test_sweep_re_embeds_when_model_id_flipped` 的 `_FakeService` 默认 vector_factory 是 4 维，新 model_id 改成 `fake-4d-fp32`（不再是 256d-fp32），并补一条 `decoded.size == 4` 断言锁住"重写后的缓存可被下一轮 validity 接受"的不变量喵～同 commit 一起吃掉了 Codex 的新一条 P1（parse_dim 锚 trailing 段）和你这条 Major（base64 validate=True + isfinite 过滤 NaN）。
